---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

GAC: Genetic Analysis of Cells
---
***

<!-- badges: start -->
[![Lifecycle: maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![R build status](https://github.com/r-lib/usethis/workflows/R-CMD-check/badge.svg)](https://github.com/r-lib/usethis/actions)
[![Codecov test coverage](https://codecov.io/gh/r-lib/usethis/branch/master/graph/badge.svg)](https://codecov.io/gh/SingerLab/gac?branch=master)
[![Travis build status](https://travis-ci.com/SingerLab/gac.svg?branch=master)](https://travis-ci.com/SingerLab/gac)
[![CRAN status](https://www.r-pkg.org/badges/version/gac)](https://CRAN.R-project.org/package=gac)
<!-- badges: end -->

# GAC is currently in ALPHA-release
R/GAC delivers am end-to-end analysis of single-cell DNA copy number by integrating quantitative genetics, statistics, and evolutionary biology.  GAC implements a simple, lightweight, and open-source R framework (Figure 1).  Inspired, but unlike Seurat and Scanpy, we integrated the logic of espressioSet/AnnData into relational matrices in native R.  This keeps the toolkit light, yet facilitates the integration of algorithms for the downstream analysis of single-cell DNA data wich is so desperately needed.  For now GAC facilitates the downstream analyses of segmented data with universal breakpoints by concurrently managing an integer copy number (X), phenotype (Y), and metadata (qc) matrices across all cells or samples.  Copy number matrices can be genrated using the output of [Varbin/Ginkgo](https://github.com/robertaboukhalil/ginkgo), [FACETS](https://github.com/mskcc/facets), [MUMdex](http://mumdex.com), [HMMcopy](https://bioconductor.org/packages/release/bioc/html/HMMcopy.html), or [SCOPE](https://github.com/rujinwang/SCOPE).  The unsegmented bin read counts is _not a correct_ input.  GAC incorporates [ape](http://ape-package.ird.fr) for phylogenetic analysis, and [ComplexHeatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/), an ultra-powerful tool for heatmaps to help visualize the data.

To implement GAC we require five easy to generate inputs:
 - a copy number / genotype matrix (X) (bins[i] x cells[j])
 - a phenotype matrix (Y) (cells[j] x phenotype[_y_])
 - a qc matrix (technical wet-lab notes) (qc) (cells[j] x qc[_c_])
 - a gene to bin index (gene.index)
 - the genomic coordinates of the bins or genotypes (chromInfo)
 - and an optional expression matrix (currently in development) (Ye; for DNA-RNA or same-cell G+T([Macaulay, et al.2015](https://dx.doi.org/10.1038/nmeth.3370))


![Figure 1.](figures/CNR.png)


## Installation

### Dependencies:
 - [ComplexHeatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/)
 - [ape](http://ape-package.ird.fr)
 - [vegan](https://github.com/vegandevs/vegan)
 - [SCclust](https://github.com/KrasnitzLab/SCclust.git)
 - [Pathview](https://pathview.uncc.edu)
 - [dplyr](https://dplyr.tidyverse.org)

```{r cran, eval = FALSE, include = FALSE, echo = FALSE}
## You can install the released version of GAC from [CRAN](https://CRAN.R-project.org) with:
# install.packages("gac")
```

You can install the development version from [GitHub](https://github.com/) with:

```{r install}
# install.packages("devtools")
devtools::install_github("SingerLab/gac")
```

## Examples
This is a basic example for drawing a copy number heatmap.  For a comprehensive overview of the package please follow the `getting_started.Rmd` in the `vignettes/`

```{r example}
library(gac)

## basic example code
data(cnr)
data(segCol)
data(legSeg)

( excl.cells <- cnr$qc$cellID[cnr$qc$qc.status == "FAIL"] )

cnr <- excludeCells(cnr, excl = excl.cells)

aH <- HeatmapCNR(cnr, what = 'X', col = segCol, show_heatmap_legend = FALSE)

draw(aH, annotation_legend_list = list(legSeg))

bH <- HeatmapCNR(cnr, what = "genes",
                 which.genes = c("CDK4", "MDM2"),
                 col = segCol, show_heatmap_legend = FALSE)

draw(bH, annotation_legend_list = list(legSeg))

```


### Motivation and design
* This package came out of the need to deliver some results.  During the 11th hour
(more like in borrowed time), I saw I was spending 85% of my time keeping 3 tables
syncornized (bins, genes, and phenotypes), 10% rendering heatmaps, and 5% actually
looking at the results.  I began to think how lucky the people who only work with 
single-cell RNAseq are to have tools like Seurat and Scanpy, how simple and 
flexible those two tools are, and how nothing for DNA copy number is as powerful
as the sister tools Seurat and Scanpy to manage the copy number matrix.  I 
eventually realized that the main diference is the restriction imposed by the
genome coordinates.  While staring at the AnnData diagram I realized that for
copy number data, the unit is a `bin` and the .X should be a matrix of common
`bins` for all cells.  However, to make biological sense of the data, **gene 
level resolution** is required.  Thus, building a syncronized matrix with genes
is of outmost importance.  At the 11th hour, having a gene to bin index
(gene.index) allowed the flexibility to interpolate the bin data to gene level
resolution and integration to the complete set of phenotypes, and QC data, but
it's not the restricted to the mouse mouse or human genomes.

The Singer Lab single-cell wet-lab and dry-lab endevours are carried forward by
a skeleton crew.  The need to have simple tools to help reduce the 85% of the 
time spent syncronizing bins, to genes, to phenotypes, and QC matrices capable
of handling a large data set of >24,000 cells was greatly needed.  Knowing the
data is growing by the week, I integrated functions to deal with the n+1 problem.  

We hope you enjoy !

- Rodrigo, et al.



# What's in the works


* Integration with MLR for non-linear genetic models

* Integration with CORE and GISTIC2 for fidning focal and recurrent events

* support for .seg files

* Cleaner code with tidyverse

* CRAN testing


# Licence
GAC framework and code is distributed under a BSD-3 License

